### ISSUE TITLE
Potential encoding issue in addition problem in play_math notebook?

### ISSUE THREAD

**Issue by ravi-annaswamy** [karpathy/minGPT#40]
Karpathy,

Thanks again for the (magic) code. Karpathy magic of simplifying the marvelous into tiny code.

I especially enjoyed the math notebook, since it opens up new applications beyond autoregressive prediction - into conditional response.

If my understanding is right, the way you have encoded the addition problem data only trains the model to learn how to predict the units digit. Please correct me if I got it wrong.

You provide x and y like this.

(tensor([4, 7, 1, 7, 0, 6]), tensor([-100, -100, -100,    0,    6,    4]))

This means you are providing the first two of the three digits in the example so the model only needs to copy them.
For instance, it may not learn how to handle carry into tens or hundreds.

The proper encoding of x that will not 'leak' the answer should be one of two ways:

1. masking the answer in the x vector:
4,7,1,7,-100,-100 -> -100, -100, -100, 0, 6, 4
This forces model to learn how to do all digits.

2. variable length example encoding with operator tokens:
<|add|>, 4, 7, <|with|>, 3, 2, 5 <|answer|>, 3, 7, 2,<|end|>
with corresponding loss mask only turned on of the last FOUR (3,7,2,<|end|>) tokens, because model needs to know how many digits to generate and stop.
(Note: I had to insert pipes in the tags to avoid git html renderer thinking they are markup for it to interpret)

There are several advantages to the second encoding:
1. you could teach add, subtract and possibly multiply.
2. the input numbers and output numbers can be variable length
3. Both these mean that we can check if model is able to generalize representation for numbers that work in addition AND subtraction and ndigits invariant math.
4. Most important you can create a long stream of inputs by placing examples back to back into a 'sentence' and use any length of input.
5. The first encoding is a seq-seq frame of mind. The second one is the GPT-way of thinking that just by predicting one more token lot can be learned. Masking only helps model not to try to predict problem itself since this formulation is a qa mode.

It is possible that my understanding may be incorrect, if so, kindly correct me.
Also I write lot of words, to articulate what is in my mind, not to mansplain, so please read with that hint!

Thanks again,
 Ravi

**Comment by fpgaminer** [comment]
GPT models use masked attention that ensure their predictions are based only on prior tokens:

https://github.com/karpathy/minGPT/blob/4050db60409b5bbaaa3302cee1e49847fc145c65/mingpt/model.py#L71

Conceptually a GPT model would be evaluated like so:

```
for i in range(len(x)):
    input = x[:i]
    target = y[i]
    output = model(input)
    loss += loss_func(output, target)
```

For example, its prediction for the 4th output token is based on the first four input tokens.  So it's trying to predict `0` based on just `4, 7, 1, 7`.

In practice though the architecture of GPT-like models allow one to compute that entire for loop in a single pass, using masking to ensure that each "column" of the model can only "see" previous tokens.

**Comment by ravi-annaswamy** [comment]
@fpgaminer Good catch! Thank you.

It was a miss on my part. I knew that GPT models use masked attention (since they are decoder-only models), but since I was processing this addition problem as a seq2seq formulation, the blindness was on me! So this is a nonissue.

Thanks for clarifying immediately.  I would like to keep the issue for a couple of hours, just for illustration to others who might have not recognized this fact, then I will close it. 

Thanks
Ravi

**Comment by ravi-annaswamy** [comment]
Non-issue, since gpt uses masked attention that always hides token from the future.
